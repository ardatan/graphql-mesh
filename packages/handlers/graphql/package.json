{
  "name": "@graphql-mesh/graphql",
  "version": "0.14.1",
  "sideEffects": false,
  "main": "dist/index.cjs.js",
  "module": "dist/index.esm.js",
  "typings": "dist/index.d.ts",
  "typescript": {
    "definition": "dist/index.d.ts"
  },
  "license": "MIT",
  "peerDependencies": {
    "graphql": "*"
  },
  "dependencies": {
    "@graphql-mesh/types": "0.35.1",
    "@graphql-mesh/utils": "0.9.0",
    "fetchache": "0.0.4",
    "@graphql-tools/url-loader": "6.8.2",
    "@graphql-tools/wrap": "7.0.5"
  },
  "publishConfig": {
    "access": "public",
    "directory": "dist"
  }
}
