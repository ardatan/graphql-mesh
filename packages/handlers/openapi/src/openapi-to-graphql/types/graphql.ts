// Copyright IBM Corp. 2018. All Rights Reserved.
// Node module: openapi-to-graphql
// This file is licensed under the MIT License.
// License text available at https://opensource.org/licenses/MIT

/**
 * Custom type definitions for GraphQL.
 */

import {
  GraphQLObjectType,
  GraphQLScalarType,
  GraphQLInputObjectType,
  GraphQLList,
  GraphQLEnumType,
  GraphQLUnionType,
  GraphQLResolveInfo,
} from 'graphql';

export enum GraphQLOperationType {
  Query,
  Mutation,
  // TODO: Subscription
}

export type GraphQLType =
  | GraphQLObjectType
  | GraphQLInputObjectType
  | GraphQLList<any>
  | GraphQLUnionType
  | GraphQLEnumType
  | GraphQLScalarType;

type Arg = {
  type: any;
  description?: string;
};

export type Args = {
  [key: string]: Arg;
};

export type ResolveFunction = (root: any, args: any, ctx: any, info: GraphQLResolveInfo) => Promise<any> | any;

export type Field = {
  type: GraphQLType;
  resolve?: ResolveFunction;
  args?: Args;
  description: string;
};
