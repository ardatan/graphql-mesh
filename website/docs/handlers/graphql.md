---
id: graphql
title: GraphQL
sidebar_label: GraphQL
---

This handler allows you to load remote GraphQL schemas and use it with schema-stitching, based on `graphql-tools-fork`.

To get started, install the handler library from NPM:

```
$ yarn add @graphql-mesh/graphql
```

Now, you can use it directly in your Mesh config file:

```yml
sources:
  - name: MyGraphQLApi
    handler:
      graphql:
        endpoint: http://my-service-url:3000/graphql
```

## Config API Reference

{@import ../generated-markdown/GraphQLHandler.generated.md}
