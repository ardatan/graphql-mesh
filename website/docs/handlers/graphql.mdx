---
id: graphql
title: GraphQL
sidebar_label: GraphQL
---

![image](https://user-images.githubusercontent.com/20847995/79219047-333ccb00-7e5a-11ea-9fce-57ff137ba924.png)

This handler allows you to load remote GraphQL schemas and use them with schema-stitching based on `graphql-tools`.

To get started, install the handler library:

<PackageInstall packages="@graphql-mesh/graphql" />

Now, you can use it directly in your Mesh config file:

```yaml
sources:
  - name: MyGraphQLApi
    handler:
      graphql:
        endpoint: http://my-service-url:3000/graphql
```

> You can check out our example that uses schema stitching with a PostgreSQL data source.
> [Click here to open the example on GitHub](https://github.com/Urigo/graphql-mesh/tree/master/examples/postgres-geodb)

<p>&nbsp;</p>

---

<p>&nbsp;</p>

## Dynamic Header Values (e.g., for Authorization)

Mesh can take dynamic values from the GraphQL Context or the environmental variables. For example, if you use `mesh dev` or `mesh start`, GraphQL Context will be the incoming HTTP request.

The expression inside dynamic values should be as in JS.

<p>&nbsp;</p>

### From Context (HTTP Header for `mesh dev` or `mesh start`)

```yaml
sources:
  - name: MyGraphQLApi
    handler:
      graphql:
        endpoint: http://my-service-url:3000/graphql
        operationHeaders:
          # Please do not use capital letters while getting the headers
          Authorization: Bearer {context.headers['x-my-api-token']}
          # You can also access to the cookies like below;
          # Authorization: Bearer {context.cookies.myApiToken}
```

And for `mesh dev` or `mesh start`, you can pass the value using `x-my-graphql-api-token` HTTP header.

<p>&nbsp;</p>

### From Environmental Variable

`MY_API_TOKEN` is the name of the environmental variable you have the value.

```yaml
sources:
  - name: MyGraphQLApi
    handler:
      graphql:
        endpoint: http://my-service-url:3000/graphql
        operationHeaders:
          Authorization: Bearer {env.MY_API_TOKEN}
```

<p>&nbsp;</p>

---

<p>&nbsp;</p>

## Config API Reference

{@import ../generated-markdown/GraphQLHandlerHTTPConfiguration.generated.md}
or
{@import ../generated-markdown/GraphQLHandlerCodeFirstConfiguration.generated.md}
