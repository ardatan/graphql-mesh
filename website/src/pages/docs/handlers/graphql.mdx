import { PackageCmd, Callout } from '@theguild/components'

# GraphQL

![image](https://user-images.githubusercontent.com/20847995/79219047-333ccb00-7e5a-11ea-9fce-57ff137ba924.png)

This handler allows you to load remote GraphQL schemas and use them with schema-stitching based on `graphql-tools`.

To get started, install the handler library:

<PackageCmd packages={['@graphql-mesh/graphql']} />

Now, you can use it directly in your Mesh config file:

```yaml
sources:
  - name: MyGraphQLApi
    handler:
      graphql:
        endpoint: http://my-service-url:3000/graphql
```

<Callout>
  You can check out our example that uses schema stitching with a PostgreSQL data source.
  [Click here to open the example on GitHub](https://github.com/Urigo/graphql-mesh/tree/master/examples/postgres-geodb)

</Callout>

## Dynamic Header Values (e.g., for Authorization)

Mesh can take dynamic values from the GraphQL Context or the environmental variables. For example, if you use `mesh dev` or `mesh start`, GraphQL Context will be the incoming HTTP request.

The expression inside dynamic values should be as in JS.

### From Context (HTTP Header for `mesh dev` or `mesh start`)

```yaml
sources:
  - name: MyGraphQLApi
    handler:
      graphql:
        endpoint: http://my-service-url:3000/graphql
        operationHeaders:
          # Please do not use capital letters while getting the headers
          # Use "{context.headers['x-my-api-token']}" if you want just the value of the header
          Authorization: Bearer {context.headers['x-my-api-token']}
          # You can also access to the cookies like below;
          # Authorization: Bearer {context.cookies.myApiToken}
```

And for `mesh dev` or `mesh start`, you can pass the value using `x-my-graphql-api-token` HTTP header.

### From Environment Variables

`MY_API_TOKEN` is the name of the environment variable you have the value.

```yaml {7}
sources:
  - name: MyGraphQLApi
    handler:
      graphql:
        endpoint: http://my-service-url:3000/graphql
        operationHeaders:
          Authorization: Bearer {env.MY_API_TOKEN}
```

## Local Schemas

We recommend providing local schema by using the `additionalTypeDefs` and `additionalResolvers` configuration options.

However, it is also possible to use a local GraphQL Schema instance as a GraphQL Mesh source, as showcased below:

```yaml
sources:
  - name: MyGraphQLApi
    handler:
      graphql:
        schema: ./my-local-schema.ts
```

```ts filename="my-local-schema.ts"
import { makeExecutableSchema } from '@graphql-tools/schema'
export default makeExecutableSchema({
  typeDefs: /* GraphQL */ `
    type Query {
      foo: String
    }
  `,
  resolvers: {
    Query: {
      foo: () => 'FOO'
    }
  }
})
```

## Fetch Strategies and Multiple HTTP endpoints for the same source

If you want to have an advanced fetch strategy for the GraphQL source such as retrying twice or timeout in 30 seconds etc.
Also, you can have different HTTP endpoints for a single source, and you can configure Mesh to get a better execution flow.

For example, you can make a request to both endpoints and return the fastest response with `race` strategy.

All `fetch` strategies can be combined to create the ultimate execution flow:

<details>
 <summary>`retry`</summary>

The `retry` mechanism allow you to specify the retry attempts for a single GraphQL endpoint/source.

The retry flow will execute in both conditions: a network error, or due to a runtime error.

```yaml {6}
sources:
  - name: uniswapv2
    handler:
      graphql:
        endpoint: https://api.thegraph.com/subgraphs/name/uniswap/uniswap-v2
        retry: 2 # specify here, if you have an unstable/error prone indexer
```

</details>

<details>
 <summary>`timeout`</summary>

The `timeout` mechanism allow you to specify the `timeout` for a given GraphQL endpoint.

```yaml {6}
sources:
  - name: uniswapv2
    handler:
      graphql:
        endpoint: https://api.thegraph.com/subgraphs/name/uniswap/uniswap-v2
        timeout: 5000 # 5 seconds
```

</details>

<details>
 <summary>`fallback`</summary>

The `fallback` mechanism allow you to specify use more than one GraphQL endpoint, for the same source.

This is helpful if you have a fallback endpoint for the same GraphQL API.

```yaml {5}
sources:
  - name: uniswapv2
    handler:
      graphql:
        strategy: fallback
        sources:
          - endpoint: https://bad-uniswap-v2-api.com
            retry: 2
            timeout: 5000
          - endpoint: https://api.thegraph.com/subgraphs/name/uniswap/uniswap-v2
```

</details>

<details>
 <summary>`race`</summary>

The `race` mechanism allow you to specify use more than one GraphQL endpoint, for the same source, and race on every execution.

If you have different places that service is deployed, this is useful to get the fastest response by racing them.

```yaml {5}
sources:
  - name: uniswapv2
    handler:
      graphql:
        strategy: race
        sources:
          - endpoint: https://bad-uniswap-v2-api.com
          - endpoint: https://api.thegraph.com/subgraphs/name/uniswap/uniswap-v2
```

</details>

## Config API Reference

import API1 from '../../../generated-markdown/GraphQLHandlerHTTPConfiguration.generated.md'
import API2 from '../../../generated-markdown/GraphQLHandlerCodeFirstConfiguration.generated.md'

<API1 />
or
<API2 />
