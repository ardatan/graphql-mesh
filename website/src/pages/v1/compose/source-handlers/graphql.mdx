---
description: GraphQL Mesh can load remote GraphQL schemas, fetch from a CDN or registry
---

import { Callout } from '@theguild/components'

# GraphQL

<Callout>
  This page is currently under construction and expected to change. Please feel free to reach out to
  us directly in case you are having any troubles.
</Callout>

This handler allows you to load remote GraphQL schemas and compose them into a supergraph.

## How to use?

If you want to load a remote GraphQL subgraph by introspecting the endpoint, you can only provide
`endpoint` in the configuration like below;

```ts filename="mesh-config.ts"
import {
  defineConfig as defineComposeConfig,
  loadGraphQLHTTPSubgraph
} from '@graphql-mesh/compose-cli'

export const composeConfig = defineComposeConfig({
  subgraphs: [
    {
      sourceHandler: loadGraphQLHTTPSubgraph('Users', {
        endpoint: 'http://localhost:4001/users'
      })
    }
  ]
})
```

### Load the schema via URL or local file (Hive CDN etc)

If you want to load a remote GraphQL subgraph from a CDN, you can provide `source` in the
configuration like below;

In this case, Mesh will take the schema from `source`, but execute the operations against the
`endpoint`.

```ts filename="mesh-config.ts" {9}
import {
  defineConfig as defineComposeConfig,
  loadGraphQLHTTPSubgraph
} from '@graphql-mesh/compose-cli'

export const composeConfig = defineComposeConfig({
  subgraphs: [
    {
      sourceHandler: loadGraphQLHTTPSubgraph('GitHub', {
        endpoint: 'https://api.github.com/graphql',
        // You can provide your SDL or introspection seperately
        source: 'https://docs.github.com/public/schema.docs.graphql',

        operationHeaders: {
          // This forwards the header from the incoming request to the remote server
          authorization: 'Bearer {context.headers["x-my-api-token"]}'
        }
      })
    }
  ]
})
```

## Headers

[Read about configuration and examples](/v1/compose/source-handlers#setting-headers)

## Fetch Strategies and Multiple HTTP endpoints for the same source

If you want to have an advanced fetch strategy for the GraphQL source such as retrying twice or
timeout in 30 seconds etc. Also, you can have different HTTP endpoints for a single source, and you
can configure Mesh to get a better execution flow.

For example, you can make a request to both endpoints and return the fastest response with `race`
strategy.

All `fetch` strategies can be combined to create the ultimate execution flow:

<details>
 <summary>`retry`</summary>

The `retry` mechanism allow you to specify the retry attempts for a single GraphQL endpoint/source.

The retry flow will execute in both conditions: a network error, or due to a runtime error.

```ts filename="mesh-config.ts" {9}
import {
  defineConfig as defineComposeConfig,
  loadGraphQLHTTPSubgraph
} from '@graphql-mesh/compose-cli'

export const composeConfig = defineComposeConfig({
  subgraphs: [
    {
      sourceHandler: loadGraphQLHTTPSubgraph('Uniswap', {
        endpoint: 'https://api.thegraph.com/subgraphs/name/uniswap/uniswap-v2',
        // Specify here, if you have an unstable/error prone indexer
        retry: 2
      })
    }
  ]
})
```

</details>

<details>
 <summary>`timeout`</summary>

The `timeout` mechanism allow you to specify the `timeout` for a given GraphQL endpoint.

```ts filename="mesh-config.ts" {9}
import {
  defineConfig as defineComposeConfig,
  loadGraphQLHTTPSubgraph
} from '@graphql-mesh/compose-cli'

export const composeConfig = defineComposeConfig({
  subgraphs: [
    {
      sourceHandler: loadGraphQLHTTPSubgraph('Uniswap', {
        endpoint: 'https://api.thegraph.com/subgraphs/name/uniswap/uniswap-v2',
        // 5 Seconds
        timeout: 5_000
      })
    }
  ]
})
```

</details>

<details>
 <summary>`fallback`</summary>

The `fallback` mechanism allow you to specify use more than one GraphQL endpoint, for the same
source.

This is helpful if you have a fallback endpoint for the same GraphQL API.

```ts filename="mesh-config.ts" {7-17}
import {
  defineConfig as defineComposeConfig,
  loadGraphQLHTTPSubgraph
} from '@graphql-mesh/compose-cli'

export const composeConfig = defineComposeConfig({
  subgraphs: [
    {
      sourceHandler: loadGraphQLHTTPSubgraph('Uniswap', {
        strategy: 'fallback',
        sources: [
          {
            endpoint: 'https://bad-uniswap-v2-api.com',
            retry: 2,
            timeout: 5_000
          },
          {
            endpoint: 'https://api.thegraph.com/subgraphs/name/uniswap/uniswap-v2'
          }
        ]
      })
    }
  ]
})
```

</details>

<details>
 <summary>`race`</summary>

The `race` mechanism allow you to specify use more than one GraphQL endpoint, for the same source,
and race on every execution.

If you have different places that service is deployed, this is useful to get the fastest response by
racing them.

```ts filename="mesh-config.ts" {7-15}
import {
  defineConfig as defineComposeConfig,
  loadGraphQLHTTPSubgraph
} from '@graphql-mesh/compose-cli'

export const composeConfig = defineComposeConfig({
  subgraphs: [
    {
      sourceHandler: loadGraphQLHTTPSubgraph('Uniswap', {
        strategy: 'race',
        sources: [
          {
            endpoint: 'https://bad-uniswap-v2-api.com'
          },
          {
            endpoint: 'https://api.thegraph.com/subgraphs/name/uniswap/uniswap-v2'
          }
        ]
      })
    }
  ]
})
```

</details>
