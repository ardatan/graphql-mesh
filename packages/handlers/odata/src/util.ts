import { ResolverData } from "@graphql-mesh/types";

export function getUrlString(url: URL) {
  return decodeURIComponent(url.toString()).split('+').join(' ');
}

export function addIdentifierToUrl(url: URL, identifierFieldName: string, identifierFieldTypeRef: string, args: any) {
  url.href += `/${args[identifierFieldName]}/`;
}
