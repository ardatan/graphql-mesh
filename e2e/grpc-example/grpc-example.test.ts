import { createTenv, type Service } from '@e2e/tenv';
import { buildHTTPExecutor } from '@graphql-tools/executor-http';
import { parse } from 'graphql';
import { inspect } from 'util';

describe('gRPC Example', () => {
  const { compose, serve, service } = createTenv(__dirname);
  let movies: Service;
  beforeAll(async () => {
    movies = await service('movies');
  });
  it('generates the correct schema', async () => {
    const { result } = await compose({ services: [movies], maskServicePorts: true });
    expect(result).toMatchSnapshot();
  });
  it('gets movies correctly', async () => {
    const { output } = await compose({ services: [movies], output: 'graphql' });
    const { execute } = await serve({ supergraph: output });
    const query = /* GraphQL */ `
      query GetMovies {
        exampleGetMovies(input: { movie: { genre: DRAMA, year: 2015 } }) {
          result {
            name
            year
            rating
            cast
            time {
              seconds
            }
          }
        }
      }
    `;
    await expect(execute({ query })).resolves.toMatchSnapshot('get-movies-grpc-example-result');
  });
  it('streams movies by cast correctly', async () => {
    const { output } = await compose({ services: [movies], output: 'graphql' });
    const { port } = await serve({ supergraph: output, pipeLogs: true });
    const executor = buildHTTPExecutor({
      endpoint: `http://localhost:${port}/graphql`,
    });
    const document = parse(/* GraphQL */ `
      query SearchMoviesByCast {
        exampleSearchMoviesByCast(input: { castName: "Tom Cruise" }) @stream {
          name
          year
          rating
          cast
        }
      }
    `);
    const result = await executor({ document });
    if (!(Symbol.asyncIterator in result)) {
      throw new Error('Expected an async iterable but received ' + inspect(result));
    }
    let i = 0;
    for await (const item of result) {
      if (item.errors?.length) {
        throw new AggregateError(item.errors, `Error in item ${i}; ${inspect(item)}`);
      }
      expect(item).toMatchSnapshot(i.toString());
      i++;
    }
  });
  it('fetches movies by cast as a subscription correctly', async () => {
    const { output } = await compose({ services: [movies], output: 'graphql' });
    const { port } = await serve({ supergraph: output });
    const executor = buildHTTPExecutor({
      endpoint: `http://localhost:${port}/graphql`,
    });
    const document = parse(/* GraphQL */ `
      subscription SearchMoviesByCast {
        exampleSearchMoviesByCast(input: { castName: "Tom Cruise" }) {
          name
          year
          rating
          cast
        }
      }
    `);
    const result = await executor({ document });
    if (!(Symbol.asyncIterator in result)) {
      throw new Error('Expected an async iterable');
    }
    let i = 0;
    for await (const item of result) {
      if (item.errors?.length) {
        throw new AggregateError(item.errors, `Error in item ${i}; ${inspect(item)}`);
      }
      expect(item).toMatchSnapshot(i.toString());
      i++;
    }
  })
});
