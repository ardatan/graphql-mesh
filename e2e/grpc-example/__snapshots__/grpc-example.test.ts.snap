// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gRPC Example generates the correct schema 1`] = `
"
schema
  @link(url: "https://specs.apollo.dev/link/v1.0")
  @link(url: "https://specs.apollo.dev/join/v0.3", for: EXECUTION)
  
  
  
  
  
  @link(
  url: "https://the-guild.dev/graphql/mesh/spec/v1.0"
  import: ["@grpcMethod", "@grpcConnectivityState", "@enum", "@grpcRootJson", "@stream", "@source", "@extraSchemaDefinitionDirective"]
)
{
  query: Query
  
  subscription: Subscription
}


  directive @join__enumValue(graph: join__Graph!) repeatable on ENUM_VALUE

  directive @join__field(
    graph: join__Graph
    requires: join__FieldSet
    provides: join__FieldSet
    type: String
    external: Boolean
    override: String
    usedOverridden: Boolean
  ) repeatable on FIELD_DEFINITION | INPUT_FIELD_DEFINITION

  directive @join__graph(name: String!, url: String!) on ENUM_VALUE

  directive @join__implements(
    graph: join__Graph!
    interface: String!
  ) repeatable on OBJECT | INTERFACE

  directive @join__type(
    graph: join__Graph!
    key: join__FieldSet
    extension: Boolean! = false
    resolvable: Boolean! = true
    isInterfaceObject: Boolean! = false
  ) repeatable on OBJECT | INTERFACE | UNION | ENUM | INPUT_OBJECT | SCALAR

  directive @join__unionMember(graph: join__Graph!, member: String!) repeatable on UNION

  scalar join__FieldSet


  directive @link(
    url: String
    as: String
    for: link__Purpose
    import: [link__Import]
  ) repeatable on SCHEMA

  scalar link__Import

  enum link__Purpose {
    """
    \`SECURITY\` features provide metadata necessary to securely resolve fields.
    """
    SECURITY

    """
    \`EXECUTION\` features provide metadata necessary for operation execution.
    """
    EXECUTION
  }







enum join__Graph {
  MOVIES @join__graph(name: "movies", url: "localhost:<movies_port>") 
}

directive @grpcMethod(
  subgraph: String
  rootJsonName: String
  objPath: String
  methodName: String
  responseStream: Boolean
) repeatable on FIELD_DEFINITION

directive @grpcConnectivityState(subgraph: String, rootJsonName: String, objPath: String)  repeatable on FIELD_DEFINITION

directive @enum(subgraph: String, value: String)  repeatable on ENUM_VALUE

directive @grpcRootJson(subgraph: String, name: String, rootJson: ObjMap, loadOptions: ObjMap)  repeatable on OBJECT

directive @stream(if: Boolean!, label: String, initialCount: Int)  on FIELD

directive @source(name: String!, type: String, subgraph: String!)  repeatable on SCALAR | OBJECT | FIELD_DEFINITION | ARGUMENT_DEFINITION | INTERFACE | UNION | ENUM | ENUM_VALUE | INPUT_OBJECT | INPUT_FIELD_DEFINITION

directive @extraSchemaDefinitionDirective(directives: _DirectiveExtensions)  repeatable on OBJECT

"""
The \`BigInt\` scalar type represents non-fractional signed whole numeric values.
"""
scalar BigInt @join__type(graph: MOVIES) 

scalar ObjMap @join__type(graph: MOVIES) 

scalar _DirectiveExtensions @join__type(graph: MOVIES) 

type Query @grpcRootJson(
  subgraph: "movies"
  name: "Root0"
  rootJson: "{\\"options\\":{\\"syntax\\":\\"proto3\\"},\\"nested\\":{\\"EmptyRequest\\":{\\"fields\\":{},\\"comment\\":null},\\"MovieRequest\\":{\\"fields\\":{\\"movie\\":{\\"type\\":\\"Movie\\",\\"id\\":1,\\"comment\\":null}},\\"comment\\":null},\\"SearchByCastRequest\\":{\\"fields\\":{\\"castName\\":{\\"type\\":\\"string\\",\\"id\\":1,\\"comment\\":null}},\\"comment\\":null},\\"MoviesResult\\":{\\"fields\\":{\\"result\\":{\\"rule\\":\\"repeated\\",\\"type\\":\\"Movie\\",\\"id\\":1,\\"comment\\":\\"list of movies\\"}},\\"comment\\":\\"movie result message, contains list of movies\\"},\\"Example\\":{\\"methods\\":{\\"GetMovies\\":{\\"requestType\\":\\"MovieRequest\\",\\"responseType\\":\\"MoviesResult\\",\\"comment\\":\\"get all movies\\"},\\"SearchMoviesByCast\\":{\\"requestType\\":\\"SearchByCastRequest\\",\\"responseType\\":\\"Movie\\",\\"responseStream\\":true,\\"comment\\":\\"search movies by the name of the cast\\"}},\\"comment\\":null},\\"AnotherExample\\":{\\"methods\\":{\\"GetMovies\\":{\\"requestType\\":\\"MovieRequest\\",\\"responseType\\":\\"MoviesResult\\",\\"comment\\":\\"get all movies\\"},\\"SearchMoviesByCast\\":{\\"requestType\\":\\"SearchByCastRequest\\",\\"responseType\\":\\"Movie\\",\\"responseStream\\":true,\\"comment\\":\\"search movies by the name of the cast\\"}},\\"comment\\":null},\\"Genre\\":{\\"values\\":{\\"UNSPECIFIED\\":0,\\"ACTION\\":1,\\"DRAMA\\":2},\\"comment\\":null,\\"comments\\":{\\"UNSPECIFIED\\":null,\\"ACTION\\":null,\\"DRAMA\\":null}},\\"Movie\\":{\\"fields\\":{\\"name\\":{\\"type\\":\\"string\\",\\"id\\":1,\\"comment\\":null},\\"year\\":{\\"type\\":\\"int64\\",\\"id\\":2,\\"comment\\":null},\\"rating\\":{\\"type\\":\\"float\\",\\"id\\":3,\\"comment\\":null},\\"cast\\":{\\"rule\\":\\"repeated\\",\\"type\\":\\"string\\",\\"id\\":4,\\"comment\\":\\"list of cast\\"},\\"time\\":{\\"type\\":\\"google.protobuf.Timestamp\\",\\"id\\":5,\\"comment\\":null},\\"genre\\":{\\"type\\":\\"Genre\\",\\"id\\":6,\\"comment\\":null}},\\"comment\\":\\"movie message payload\\"},\\"google\\":{\\"nested\\":{\\"protobuf\\":{\\"nested\\":{\\"Timestamp\\":{\\"fields\\":{\\"seconds\\":{\\"type\\":\\"int64\\",\\"id\\":1},\\"nanos\\":{\\"type\\":\\"int32\\",\\"id\\":2}},\\"comment\\":null}}}}}}}"
) @extraSchemaDefinitionDirective(
  directives: {transport: [{subgraph: "movies", kind: "grpc", location: "localhost:<movies_port>", options: {requestTimeout: 200000, metaData: {someKey: "someValue", connection_type: "{context.headers.connection}"}}}]}
) @join__type(graph: MOVIES)  {
  """
  get all movies
  """
  exampleGetMovies(input: MovieRequest_Input) : MoviesResult @grpcMethod(
    subgraph: "movies"
    rootJsonName: "Root0"
    objPath: "Example"
    methodName: "GetMovies"
    responseStream: false
  ) @source(name: "Example_GetMovies", type: "MoviesResult", subgraph: "movies") 
  """
  search movies by the name of the cast
  """
  exampleSearchMoviesByCast(input: SearchByCastRequest_Input) : [Movie] @grpcMethod(
    subgraph: "movies"
    rootJsonName: "Root0"
    objPath: "Example"
    methodName: "SearchMoviesByCast"
    responseStream: true
  ) @source(name: "Example_SearchMoviesByCast", type: "[Movie]", subgraph: "movies") 
  exampleConnectivityState(tryToConnect: Boolean) : ConnectivityState @grpcConnectivityState(subgraph: "movies", rootJsonName: "Root0", objPath: "Example")  @source(
    name: "Example_connectivityState"
    type: "ConnectivityState"
    subgraph: "movies"
  )
  """
  get all movies
  """
  anotherExampleGetMovies(input: MovieRequest_Input) : MoviesResult @grpcMethod(
    subgraph: "movies"
    rootJsonName: "Root0"
    objPath: "AnotherExample"
    methodName: "GetMovies"
    responseStream: false
  ) @source(name: "AnotherExample_GetMovies", type: "MoviesResult", subgraph: "movies") 
  """
  search movies by the name of the cast
  """
  anotherExampleSearchMoviesByCast(input: SearchByCastRequest_Input) : [Movie] @grpcMethod(
    subgraph: "movies"
    rootJsonName: "Root0"
    objPath: "AnotherExample"
    methodName: "SearchMoviesByCast"
    responseStream: true
  ) @source(
    name: "AnotherExample_SearchMoviesByCast"
    type: "[Movie]"
    subgraph: "movies"
  )
  anotherExampleConnectivityState(tryToConnect: Boolean) : ConnectivityState @grpcConnectivityState(subgraph: "movies", rootJsonName: "Root0", objPath: "AnotherExample")  @source(
    name: "AnotherExample_connectivityState"
    type: "ConnectivityState"
    subgraph: "movies"
  )
}

"""
movie result message, contains list of movies
"""
type MoviesResult @join__type(graph: MOVIES)  {
  """
  list of movies
  """
  result: [Movie]
}

"""
movie message payload
"""
type Movie @join__type(graph: MOVIES)  {
  name: String
  year: BigInt
  rating: Float
  """
  list of cast
  """
  cast: [String]
  time: google__protobuf__Timestamp
  genre: Genre
}

type google__protobuf__Timestamp @join__type(graph: MOVIES)  {
  seconds: BigInt
  nanos: Int
}

type Subscription @join__type(graph: MOVIES)  {
  """
  search movies by the name of the cast
  """
  exampleSearchMoviesByCast(input: SearchByCastRequest_Input) : Movie @grpcMethod(
    subgraph: "movies"
    rootJsonName: "Root0"
    objPath: "Example"
    methodName: "SearchMoviesByCast"
    responseStream: true
  ) @source(name: "Example_SearchMoviesByCast", type: "Movie", subgraph: "movies") 
  """
  search movies by the name of the cast
  """
  anotherExampleSearchMoviesByCast(input: SearchByCastRequest_Input) : Movie @grpcMethod(
    subgraph: "movies"
    rootJsonName: "Root0"
    objPath: "AnotherExample"
    methodName: "SearchMoviesByCast"
    responseStream: true
  ) @source(name: "AnotherExample_SearchMoviesByCast", type: "Movie", subgraph: "movies") 
}

enum ConnectivityState @join__type(graph: MOVIES)  {
  IDLE @join__enumValue(graph: MOVIES) 
  CONNECTING @join__enumValue(graph: MOVIES) 
  READY @join__enumValue(graph: MOVIES) 
  TRANSIENT_FAILURE @join__enumValue(graph: MOVIES) 
  SHUTDOWN @join__enumValue(graph: MOVIES) 
}

enum Genre @join__type(graph: MOVIES)  {
  UNSPECIFIED @enum(subgraph: "movies", value: "0")  @join__enumValue(graph: MOVIES) 
  ACTION @enum(subgraph: "movies", value: "1")  @join__enumValue(graph: MOVIES) 
  DRAMA @enum(subgraph: "movies", value: "2")  @join__enumValue(graph: MOVIES) 
}

input MovieRequest_Input @join__type(graph: MOVIES)  {
  movie: Movie_Input
}

"""
movie message payload
"""
input Movie_Input @join__type(graph: MOVIES)  {
  name: String
  year: BigInt
  rating: Float
  """
  list of cast
  """
  cast: [String]
  time: google__protobuf__Timestamp_Input
  genre: Genre
}

input google__protobuf__Timestamp_Input @join__type(graph: MOVIES)  {
  seconds: BigInt
  nanos: Int
}

input SearchByCastRequest_Input @join__type(graph: MOVIES)  {
  castName: String
}
    
"
`;

exports[`gRPC Example gets movies correctly: get-movies-grpc-example-result 1`] = `
{
  "data": {
    "exampleGetMovies": {
      "result": [
        {
          "cast": [
            "Leonardo DiCaprio",
            "Jonah Hill",
            "Margot Robbie",
          ],
          "name": "The Wolf of Wall Street",
          "rating": 0.7799999713897705,
          "time": {
            "seconds": 1608422400000,
          },
          "year": 0,
        },
      ],
    },
  },
}
`;

exports[`gRPC Example streams movies by cast correctly: 0 1`] = `
{
  "data": {
    "exampleSearchMoviesByCast": [],
  },
}
`;

exports[`gRPC Example streams movies by cast correctly: 1 1`] = `
{
  "data": {
    "exampleSearchMoviesByCast": [
      {
        "cast": [
          "Tom Cruise",
          "Simon Pegg",
          "Jeremy Renner",
        ],
        "name": "Mission: Impossible Rogue Nation",
        "rating": 0.9700000286102295,
        "year": 0,
      },
    ],
  },
}
`;

exports[`gRPC Example streams movies by cast correctly: 2 1`] = `
{
  "data": {
    "exampleSearchMoviesByCast": [
      {
        "cast": [
          "Tom Cruise",
          "Simon Pegg",
          "Jeremy Renner",
        ],
        "name": "Mission: Impossible Rogue Nation",
        "rating": 0.9700000286102295,
        "year": 0,
      },
      {
        "cast": [
          "Tom Cruise",
          "Simon Pegg",
          "Henry Cavill",
        ],
        "name": "Mission: Impossible - Fallout",
        "rating": 0.9300000071525574,
        "year": 0,
      },
    ],
  },
}
`;

exports[`gRPC Example streams movies by cast correctly: 3 1`] = `
{
  "data": {
    "exampleSearchMoviesByCast": [
      {
        "cast": [
          "Tom Cruise",
          "Simon Pegg",
          "Jeremy Renner",
        ],
        "name": "Mission: Impossible Rogue Nation",
        "rating": 0.9700000286102295,
        "year": 0,
      },
      {
        "cast": [
          "Tom Cruise",
          "Simon Pegg",
          "Henry Cavill",
        ],
        "name": "Mission: Impossible - Fallout",
        "rating": 0.9300000071525574,
        "year": 0,
      },
      null,
    ],
  },
}
`;

exports[`gRPC Example streams movies by cast correctly: 4 1`] = `
{
  "data": {
    "exampleSearchMoviesByCast": [
      {
        "cast": [
          "Tom Cruise",
          "Simon Pegg",
          "Jeremy Renner",
        ],
        "name": "Mission: Impossible Rogue Nation",
        "rating": 0.9700000286102295,
        "year": 0,
      },
      {
        "cast": [
          "Tom Cruise",
          "Simon Pegg",
          "Henry Cavill",
        ],
        "name": "Mission: Impossible - Fallout",
        "rating": 0.9300000071525574,
        "year": 0,
      },
      null,
    ],
  },
}
`;
